syntax = "proto3";

package examples.sensor;

import "litepb/rpc_options.proto";

enum SensorStatus {
    OK = 0;
    WARNING = 1;
    ERROR = 2;
}

message ReadingRequest {
    int32 sensor_id = 1;
}

message ReadingResponse {
    int32 sensor_id = 1;
    float temperature = 2;
    SensorStatus status = 3;
}

message AlertEvent {
    int32 sensor_id = 1;
    float temperature = 2;
    SensorStatus status = 3;
    string message = 4;
}

message AlertAck {
    bool received = 1;
}

message AlertNotification {
    int32 sensor_id = 1;
    float temperature = 2;
    SensorStatus status = 3;
    string message = 4;
}

service SensorService {
    option (litepb.rpc.service_id) = 1;
    
    rpc GetReading(ReadingRequest) returns (ReadingResponse) {
        option (litepb.rpc.method_id) = 1;
    }
    rpc NotifyAlert(AlertEvent) returns (AlertAck) {
        option (litepb.rpc.method_id) = 2;
    }
    rpc SendAlert(AlertNotification) returns (AlertNotification) {
        option (litepb.rpc.method_id) = 3;
        option (litepb.rpc.fire_and_forget) = true;
    }
}
